kubernetes-crds/pkg/apis/projectriff.io/v1/function_invoker.go

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1
+
+import (
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:defaulter-gen=true
+
+// Represents the functions.projectriff.io CRD
+type FunctionInvoker struct {
+	meta_v1.TypeMeta   `json:",inline"`
+	meta_v1.ObjectMeta `json:"metadata"`
+	Spec               FunctionInvokerSpec   `json:"spec"`
+	Status             FunctionInvokerStatus `json:"status,omitempty"`
+}
+
+// Spec (what the user wants) for a function invoker.
+type FunctionInvokerSpec struct {
+
+	// Invoker version
+	Version string `json:"version"`
+
+	// File patterns to match this invoker with for an artifact
+	Matchers []string `json:"matchers"`
+
+	// Properties for the invoker
+	// +optional
+	Properties FunctionInvokerProperties `json:"properties,omitifempty"`
+
+	// Files to generate for this invoker
+	Files []FunctionInvokerFile `json:"files"`
+}
+
+// Properties (invoker properties) for a function invoker
+type FunctionInvokerProperties struct {
+
+	// If true, the invoker requires a handler function name
+	// +optional
+	HandlerRequired bool `json:"handler,omitifempty"`
+
+	// Default function properties
+	// +optional
+	Function FunctionSpec `json:"function,omitifempty"`
+
+	// Default topic properties
+	// +optional
+	Topics TopicSpec `json:"topics,omitifempty"`
+}
+
+// File (generated files) for a function invoker
+type FunctionInvokerFile struct {
+
+	// File name to generate
+	Name string `json:"name"`
```

Comment:
Maybe this would be better named `Path`. I believe it could also require dynamic parts in it, so could be a template itself

Reply:
now Path, holding off on making it templated.